### ISSUE TITLE
twisted/conch/test/test_recvline.py stdio loopback tests set up child environment in a problematic way

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4203]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#4203|
|Type|defect|
|Created|2010-01-10 17:00:01Z|
|Branch|https://github.com/twisted/twisted/tree/full-child-environment-4203|

`_StdioMixin.setUp` launches a child process for the tests to interact with.  It creates the child with a very minimal environment, though.  This causes issues for PyPy, which relies on ctypes to provide certain modules; ctypes, in turn, requires that `objdump` be on the `PATH` and fails grossly when it is not.

More generally, it's a bit optimistic to expect that semi-arbitrary programs (eg !CPython or PyPy) beyond our control will work properly with such a minimal environment.  Reproducing the entire parent environment is probably a better idea.

<details><summary>Searchable metadata</summary>

```
trac-id__4203 4203
type__defect defect
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_full_child_environment_4203 branches/full-child-environment-4203
branch_author__ 
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__ 
time__1263142801000000 1263142801000000
changetime__1263891866000000 1263891866000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [27945]) Copy the parent environment to the child, plus the PYTHONPATH modification

refs [#4203](https://github.com/twisted/twisted/issues/4203)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [27947]) Change determineDefaultFunctionName to look up from the end instead of assuming it can start at the beginning

refs [#4203](https://github.com/twisted/twisted/issues/4203)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

(In [27948]) Raise a custom exception instead of triggering a ZeroDivsionError

refs [#4203](https://github.com/twisted/twisted/issues/4203)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun removed owner|
|-|-|

I ended up making three changes (the ones mentioned above).  The first should be obvious from the ticket description.  The second is a fix to some nearby code which I took this opportunity to fix (it's a bit unrelated, but I'm not all that worried about this branch being too big).  It changes some test helper code to pick up the correct string to use when comparing results.  Previously it picked up the wrong string on PyPy because it assumed the stack started with the module scope frame, but on PyPy there are some extra frames above that.  Likewise, the third changes the exception raised in one test so that it does not rely on the exact string representation of a ZeroDivisionError raised by the interpreter.  All three of these changes are purely test fixes.

A test run from trunk on a recent version of PyPy results in 40 failures/errors (<http://buildbot.twistedmatrix.com/builders/ubuntu32-pypy/builds/37>) after almost 30 minutes (mainly due to tests which hang until the trial timeout kicks in).  A test run of this branch produces only 27 failures/errors (<http://buildbot.twistedmatrix.com/builders/ubuntu32-pypy/builds/39>) and takes only about five minutes.

The tests also all pass on all the supported platforms (<http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/full-child-environment-4203>).

**Comment by twisted-trac** [comment]
|[<img alt="moshez's avatar" src="https://avatars.githubusercontent.com/u/422703?s=50" width="50" height="50">](https://github.com/moshez)<a name="note_5"></a>|@moshez set owner to @exarkun|
|-|-|

This seems fine to me.

This should be checked-in as-is.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set status to `closed`|
|-|-|

(In [28027]) Merge full-child-environment-4203

Author: exarkun
Reviewer: moshez
Fixes: [#4203](https://github.com/twisted/twisted/issues/4203)

For the Twisted Conch Insults tests which interact with a child process, ensure that
the child process is set up with an environment more closely resembling the parent,
in case any of its contents are necessary for proper operation.  In particular, this
fixes these tests on PyPy, which currently requires objdump to be somewhere on PATH.

Also account for some minor differences in the way certain exceptions behave between
CPython and PyPy in these tests.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
